src/desktop/credentials/DesktopNativeCredentialsFacade.ts

```diff
@@ -1,35 +1,136 @@
+		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
+		data = this.crypto.aes256EncryptKey(credentialsKey, data)
+		data = await this.maybeEncryptUsingAppPin(data, encryptionMode)
+		return data
 	}
 
-	async encryptUsingKeychain(data: Uint8Array, encryptionMode: CredentialEncryptionMode.DEVICE_LOCK): Promise<Uint8Array> {
-		// making extra sure that the mode is the right one since this comes over IPC
-		assert(encryptionMode === CredentialEncryptionMode.DEVICE_LOCK, "should not use unsupported encryption mode")
-		const key = await this._desktopKeyStoreFacade.getCredentialsKey()
-		const encryptedData = this._crypto.aes256EncryptKey(key, data)
-		return Promise.resolve(encryptedData)
+	private async maybeDecryptUsingAppPin(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		if (encryptionMode === CredentialEncryptionMode.APP_PIN) {
+			const appPinKey = await this.resolveKeyFromAppPin(false)
+			if (appPinKey == null) {
+				// there is no salt, so someone might have turned off app pin in the meantime.
+				// if we actually have pin-encrypted credentials and just lost the salt, we
+				// will get a decryption failure and delete the credentials further up the stack.
+				return data
+			}
+			try {
+				data = this.crypto.aesDecryptBytes(appPinKey, data)
+			} catch (e) {
+				if (e instanceof CryptoError) {
+					const nativeFacade = await this.getCurrentCommonNativeFacade()
+					nativeFacade.showAlertDialog("invalidPassword_msg")
+					throw new CancelledError("app pin verification failed")
+				} else {
+					throw e
+				}
+			}
+			return data
+		} else {
+			// our mode is not app pin, so the app pin should not be set.
+			// this is technically not required because we're deleting it when
+			// encrypting with another mode, but things might have gone wrong.
+			await this.conf.setVar(DesktopConfigKey.appPinSalt, null)
+			return data
+		}
+	}
+
+	private async maybeEncryptUsingAppPin(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		if (encryptionMode === CredentialEncryptionMode.APP_PIN) {
+			const appPinKey = await this.resolveKeyFromAppPin(true)
+			if (appPinKey == null) {
+				throw new ProgrammingError("we should have generated a new key or thrown a CancelledErrorq")
+			}
+			data = this.crypto.aesEncryptBytes(appPinKey, data)
+			return data
+		} else {
+			// our mode is not app pin, so the app pin salt should not be set
+			await this.conf.setVar(DesktopConfigKey.appPinSalt, null)
+			return data
+		}
+	}
+
+	/**
+	 * if there is a salt stored, use it and a password prompt to derive the app pin key.
+	 * if there isn't, ask for a new password, generate a salt & store it, then derive the key.
+	 * @param generate whether to generate a new key if no salt is found
+	 * @return the derived 256 bit key or null if none is found and generate is false
+	 */
+	private async resolveKeyFromAppPin(generate: boolean): Promise<Aes256Key | null> {
```

Comment:
I know this mirrors our existing `KeyStoreFacade` behavior but I just want to point out that I think it's not a good idea to try to always give the key, because we run into fun bugs where the key disappears and we generate a new one and it doesn't match anymore. It should ideally be two methods and in "set pin" case we would do `getKey() ?? generateAndStoreKey()`.

The reason why it was a single method in other cases was probably 1. pre licc so each additional method is pain 2. avoiding authenticating user twice (I am not even sure if this would be the case)